elide-graphql/src/main/java/com/yahoo/elide/graphql/ModelBuilder.java

```diff
@@ -150,12 +152,63 @@ public ModelBuilder(EntityDictionary entityDictionary,
         queryObjectRegistry = new HashMap<>();
         connectionObjectRegistry = new HashMap<>();
         excludedEntities = new HashSet<>();
+        relationshipOpArgument = new HashMap<>();
     }
 
     public void withExcludedEntities(Set<Type<?>> excludedEntities) {
         this.excludedEntities = excludedEntities;
     }
 
+    public GraphQLArgument getRelationshipOp(Type<?> clazz) {
+        GraphQLArgument existing = relationshipOpArgument.get(clazz);
+        if (existing != null) {
+            return existing;
+        }
+
+        String entityName = entityDictionary.getJsonAliasFor(clazz);
```

Comment:
I'm really confused what this code is doing.  It looks like it is changing the value of relationship op to include the entity name.  I'm not following the intent here, but we don't want to change the api contract.   I also feel like we really need two distinct methods here - one to get the relationship ops for a root level entity and one to get the relationship ops for a relationship.

Reply:
I'm actually not that clear on the logic for determining the relationship ops from the permissions for graphql. Is there a difference in allowable values in the relationship ops for the root entity vs the relationship?

Reply:
Permissions are not as straightforward given that graphql mutations behave more like patch extension requests in JSON-API.  Whether you can perform the operation or not depends not just on the operation being performed but on the data passed along with the operation.

Also some operations like UPSERT and REPLACE are really doing different things depending on the data.

Root entities:

- UPSERT - CreatePermission OR UpdatePermission (depending on whether the object already exists or not)
- DELETE - DeletePermission
- REMOVE - DeletePermission
- REPLACE - CreatePermission OR UpdatePermission OR DeletePermission
- FETCH - ReadPermission
- UPDATE - UpdatePermission

Relationships:
- FETCH - ReadPermission (on the field)
- REMOVE - UpdatePermission (on the field)
- DELETE - DeletePermission (on the type being deleted)
- UPSERT - CreatePermission (on the type being created) OR UpdatePermission(on the field)
- REPLACE - CreatePermission OR UpdatePermission (field and type being manipulated) OR DeletePermission
- UPDATE - UpdatePermission (on the field)

We can problem simplify this though to the following:
Root entities:

- UPSERT - CreatePermission
- DELETE - DeletePermission
- REMOVE - DeletePermission
- REPLACE - Too complex to put a restriction on.  Just always allow it.
- FETCH - ReadPermission
- UPDATE - UpdatePermission

Relationships:
- FETCH - ReadPermission (on the field)
- REMOVE - UpdatePermission (on the field)
- DELETE - DeletePermission (on the type being deleted)
- UPSERT - UpdatePermission (on the field)
- REPLACE - UpdatePermission (on the field)
- UPDATE - UpdatePermission (on the field)

Alternatively, we don't solve for GraphQL at all this way.

Reply:
I think it makes sense to at least attempt to filter the operations as that is better than nothing. I retained some criteria of restricting REPLACE as I think the read only use case should be relatively common and exposing REPLACE is misleading.